components/plan-build/bin/composite_build_functions.sh

```diff
@@ -152,10 +152,21 @@ _validate_bind_mappings() {
 
     resolved="${resolved_services[$pkg]}"
 
-    while read -r line; do
-      IFS== read bind_name exports <<< "${line}"
-      all_binds_for_pkg[$bind_name]="${exports[@]}"
-    done < <(_read_metadata_file_for "${resolved}" BINDS)
+    if [[ -f "${resolved}/BINDS" ]]; then
+      while read -r line; do
+        IFS== read bind_name exports <<< "${line}"
+        all_binds_for_pkg[$bind_name]="${exports[@]}"
+      done < <(_read_metadata_file_for "${resolved}" BINDS)
+    fi
+    if [[ -f "${resolved}/BINDS_OPTIONAL" ]]; then
+      while read -r line; do
+        IFS== read bind_name exports <<< "${line}"
+        if [[ -n "${all_binds_for_pkg[$bind_name]}" ]]; then
+          exit_with "The bind ${bind_name} has already been declared in pkg_binds for package ${pkg}, it cannot also be declared in pkg_binds_optional"
+        fi
+        all_binds_for_pkg[$bind_name]="${exports[@]}"
+      done < <(_read_metadata_file_for "${resolved}" BINDS_OPTIONAL)
+    fi
```

Comment:
Ah, I missed this earlier... `_read_metadata_file_for` includes a file existence check, so you don't need to do those here. You can just add a while loop for `BINDS_OPTIONAL` that looks like the original one for `BINDS`.

After that, we can get this merged.

Reply:
hm so the return behavior of `_read_metadata_file_for` is problematic here, because we try to do something with that empty line:

```
++ [[ -f /hab/pkgs/chef-server/chef-server-nginx/0.1.0/20171116175937/BINDS ]]
++ echo
+ IFS==
+ read bind_name exports
+ all_binds_for_pkg[$bind_name]=
/hab/pkgs/core/hab-plan-build/0.39.1/20171118010547/bin/composite_build_functions.sh: line 157: all_binds_for_pkg[$bind_name]: bad array subscript
```

commenting out the `else echo` fixes that.   WDYT?